x/gamm/types/pool.go

```diff
@@ -75,15 +75,26 @@ type PoolI interface {
 	PokePool(blockTime time.Time)
 }
 
-// PoolExitSwapExactAmountOutExtension is an extension of the PoolI
+// PoolAmountOutExtension is an extension of the PoolI
 // interface definiting an abstraction for pools that hold tokens.
-// In addition, it supports ExitSwapExactAmountOut method.
-// See definition below.
-type PoolExitSwapExactAmountOutExtension interface {
+// In addition, it supports JoinSwapShareAmountOut and ExitSwapExactAmountOut methods
+// that allow joining with the exact amount of shares to get out, and exiting with exact
+// amount of coins to get out.
+// See definitions below.
+type PoolAmountOutExtension interface {
 	PoolI
 
+	// JoinSwapShareAmountOut add liquidity to a specified pool with a maximum amount of tokens in (tokenInMaxAmount)
+	// and swaps to an exact number of shares (shareOutAmount).
+	JoinSwapShareAmountOut(
+		ctx sdk.Context,
+		tokenInDenom string,
+		shareOutAmount sdk.Int,
+		tokenInMaxAmount sdk.Int,
```

Comment:
I suggest we remove `tokenInMaxAmount` as an argument here

Reply:
Yeah I think so. Ideally, slippage bounds checks are something that the keeper layer is responsible for handling, so we know its done correctly independent of the underlying pools implementation.

Reply:
It seems that I've encountered a similar rounding error to #1196. More investigation is needed to determine if these are related but I'm just documenting the findings for now.

The issue here is related to the panic in the code below:
```
func (k Keeper) JoinSwapShareAmountOut(
	ctx sdk.Context,
	sender sdk.AccAddress,
	poolId uint64,
	tokenInDenom string,
	shareOutAmount sdk.Int,
	tokenInMaxAmount sdk.Int,
) (tokenInAmount sdk.Int, err error) {
	pool, err := k.getPoolForSwap(ctx, poolId)
	if err != nil {
		return sdk.Int{}, err
	}

	extendedPool, ok := pool.(types.PoolAmountOutExtension)
	if !ok {
		return sdk.Int{}, fmt.Errorf("pool with id %d does not support this kind of join", poolId)
	}

	tokenInAmount, err = extendedPool.CalcTokenInShareAmountOut(ctx, tokenInDenom, shareOutAmount, pool.GetSwapFee(ctx))
	if err != nil {
		return sdk.Int{}, err
	}

	if tokenInAmount.GT(tokenInMaxAmount) {
		return sdk.Int{}, sdkerrors.Wrapf(types.ErrLimitMaxAmount, "%d resulted tokens is larger than the max amount of %d", tokenInAmount.Int64(), tokenInMaxAmount.Int64())
	}

	tokenIn := sdk.Coins{sdk.NewCoin(tokenInDenom, tokenInAmount)}

	sharesCreated, err := extendedPool.JoinPool(ctx, tokenIn, pool.GetSwapFee(ctx))
	if sharesCreated != shareOutAmount {
		panic(fmt.Sprintf("%d shares created is not equal to the requested amount of %d", sharesCreated, shareOutAmount))
	}

	err = k.applyJoinPoolStateChange(ctx, pool, sender, shareOutAmount, tokenIn)
	if err != nil {
		return sdk.ZeroInt(), err
	}
	return tokenInAmount, nil
}
```

As per suggestion, I attempted to move validation logic to the keeper layer by refactoring to create a non-mutative method `CalcTokenInShareAmountOut`. Then, I noticed that it should be possible to reuse the existing `JoinPool` of the `PoolI` interface after validation on the keeper layer.

The resulting shares, returned from `JoinPool`, should be equal to the requested `shareOutAmount`. However, the above functionality panics with:
```
test panicked: {824636220864} shares created is not equal to the requested amount of {824633847808}
```

I will investigate further if these are related with #1196  or if this is an issue specific to my implementation

Reply:
Thanks, yeah this does look something has gone wrong

Reply:
Oh wow, surprised that we had such rounding issues! Thanks for investigating on this by comparing it with v7.x!